File: scripts/fetch_usda.py
import urllib.request, urllib.parse, json, time, sys

API_KEY = "DEMO_KEY"
DATA_TYPE = "SR Legacy"

def fetch_usda(query):
    url = f"https://api.nal.usda.gov/fdc/v1/foods/search?api_key={API_KEY}&query={urllib.parse.quote(query)}&pageSize=1&dataType={DATA_TYPE}"
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "DietTracker/1.0"})
        with urllib.request.urlopen(req, timeout=15) as resp:
            data = json.loads(resp.read())
        food = data.get("foods", [None])[0]
        if not food:
            return None
        nutrients = {}
        for n in food.get("foodNutrients", []):
            name = n.get("nutrientName", "")
            if "Energy" in name: nutrients["energy"] = n["value"]
            if name == "Protein": nutrients["protein"] = n["value"]
            if "Total lipid (fat)" in name: nutrients["fat"] = n["value"]
            if "Carbohydrate" in name: nutrients["carbs"] = n["value"]
        if not nutrients.get("energy"):
            return None
        return {
            "usda_name": food.get("description", ""),
            "energy": round(nutrients.get("energy", 0), 1),
            "protein": round(nutrients.get("protein", 0), 1),
            "fat": round(nutrients.get("fat", 0), 1),
            "carbs": round(nutrients.get("carbs", 0), 1),
        }
    except Exception as e:
        return {"error": str(e)}

FOODS = [
    ("Rice, white, cooked, no salt", "쌀밥", "korean", "rice"),
    ("Chicken, breast, roasted, skinless", "닭가슴살", "korean", "meat"),
    ("Egg, whole, cooked, fried", "계란후라이", "korean", "side"),
    ("Beef, ground, 85% lean, broiled", "소고기다짐육", "korean", "meat"),
    ("Salmon, Atlantic, farmed, cooked", "연어구이", "korean", "fish"),
    ("Tofu, raw, firm, prepared with calcium sulfate", "두부", "korean", "side"),
    ("Pork, fresh, belly, raw", "삼겹살", "korean", "meat"),
    ("Noodles, egg, cooked, enriched", "파스타면", "western", "noodle"),
    ("Bread, wheat", "식빵", "western", "bread"),
    ("Potatoes, french fried, salt added, frozen", "감자튀김", "western", "side"),
    ("Bacon, cooked", "베이컨", "western", "meat"),
    ("Bagels, plain, enriched", "베이글", "western", "bread"),
    ("Pizza, cheese, regular", "치즈피자", "western", "bread"),
    ("Cheese, cheddar", "체다치즈", "western", "side"),
    ("Yogurt, plain, low fat", "플레인요거트", "western", "side"),
    ("Milk, whole, 3.25% milkfat", "우유", "western", "side"),
    ("Bananas, raw", "바나나", "western", "side"),
    ("Apples, raw", "사과", "western", "side"),
    ("Sweet potato, cooked, baked in skin", "고구마", "korean", "side"),
    ("Corn, sweet, yellow, cooked, boiled", "옥수수", "korean", "side"),
    ("Potatoes, boiled, cooked without skin, flesh", "감자", "korean", "side"),
    ("Onions, raw", "양파", "korean", "side"),
    ("Garlic, raw", "마늘", "korean", "side"),
    ("Carrots, raw", "당근", "korean", "side"),
    ("Spinach, raw", "시금치", "korean", "side"),
    ("Cabbage, raw", "양배추", "korean", "side"),
    ("Butter, salted", "버터", "western", "side"),
    ("Oil, olive, salad or cooking", "올리브유", "western", "side"),
    ("Honey", "꿀", "western", "side"),
    ("Sugars, granulated", "설탕", "western", "side"),
    ("Ice creams, vanilla", "바닐라아이스크림", "western", "snack"),
    ("Peanut butter, smooth style", "땅콩버터", "western", "side"),
]

results = []
errors = []
for query, name, cat, subcat in FOODS:
    sys.stderr.write(f"Fetching {name}... ")
    sys.stderr.flush()
    data = fetch_usda(query)
    if data and "error" not in data:
        results.append({
            "name": name, "category": cat, "subcategory": subcat,
            "caloriesPer100g": data["energy"],
            "proteinPer100g": data["protein"],
            "fatPer100g": data["fat"],
            "carbsPer100g": data["carbs"],
            "source": f"USDA: {data['usda_name']}",
        })
        sys.stderr.write(f"✅ {data['energy']}kcal\n")
    else:
        errors.append(name)
        sys.stderr.write(f"❌ {data.get('error', 'no data') if isinstance(data, dict) else 'no data'}\n")
    time.sleep(0.3)  # Rate limiting

print(json.dumps(results, indent=2, ensure_ascii=False))
sys.stderr.write(f"\n✅ {len(results)}/{len(FOODS)} fetched, ❌ {len(errors)} failed\n")
if errors:
    sys.stderr.write(f"Failed: {errors}\n")

File: scripts/fetch-usda.js
// USDA FoodData Central fetcher
// Uses USDA demo API key - limited to 30 requests/hour
const https = require('https');
const API_KEY = 'DEMO_KEY';

async function fetchUSDA(query, dataType = 'SR Legacy') {
  return new Promise((resolve) => {
    const url = `https://api.nal.usda.gov/fdc/v1/foods/search?api_key=${API_KEY}&query=${encodeURIComponent(query)}&pageSize=1&dataType=${encodeURIComponent(dataType)}`;
    https.get(url, (res) => {
      let data = '';
      res.on('data', c => data += c);
      res.on('end', () => {
        try {
          const j = JSON.parse(data);
          const f = j.foods?.[0];
          if (!f) { resolve(null); return; }
          const n = {};
          (f.foodNutrients || []).forEach(nut => {
            if (nut.nutrientName === 'Energy') n.energy = nut.value;
            if (nut.nutrientName === 'Protein') n.protein = nut.value;
            if (nut.nutrientName === 'Total lipid (fat)') n.fat = nut.value;
            if (nut.nutrientName === 'Carbohydrate, by difference') n.carbs = nut.value;
          });
          resolve({ name: f.description, ...n });
        } catch { resolve(null); }
      });
    }).on('error', () => resolve(null));
  });
}

const FOODS = [
  { query: 'Rice, white, cooked', name: '쌀밥', category: 'korean', subcategory: 'rice', servings: [{unitName:'공기',gramsPerUnit:210},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Rice, brown, cooked', name: '현미밥', category: 'korean', subcategory: 'rice', servings: [{unitName:'공기',gramsPerUnit:210},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Egg, whole, cooked, fried', name: '계란후라이', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:55},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Chicken, breast, roasted', name: '닭가슴살', category: 'korean', subcategory: 'meat', servings: [{unitName:'인분',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Tofu, firm', name: '두부', category: 'korean', subcategory: 'side', servings: [{unitName:'모',gramsPerUnit:300},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Beef, short loin, top loin, steak, separable lean only, broiled', name: '스테이크', category: 'western', subcategory: 'meat', servings: [{unitName:'인분',gramsPerUnit:250},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Pork, fresh, belly, raw', name: '삼겹살', category: 'korean', subcategory: 'meat', servings: [{unitName:'인분',gramsPerUnit:200},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Salmon, Atlantic, farmed, raw', name: '연어', category: 'korean', subcategory: 'fish', servings: [{unitName:'인분',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Noodles, egg, cooked, enriched', name: '파스타면', category: 'western', subcategory: 'noodle', servings: [{unitName:'접시',gramsPerUnit:200},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Bread, wheat', name: '식빵', category: 'western', subcategory: 'bread', servings: [{unitName:'조각',gramsPerUnit:30},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Potatoes, french fried, all types, salt added', name: '감자튀김', category: 'western', subcategory: 'side', servings: [{unitName:'접시',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Bacon, cooked', name: '베이컨', category: 'western', subcategory: 'meat', servings: [{unitName:'조각',gramsPerUnit:20},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Bagel, plain, enriched', name: '베이글', category: 'western', subcategory: 'bread', servings: [{unitName:'개',gramsPerUnit:100},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Pizza, cheese', name: '치즈피자', category: 'western', subcategory: 'bread', servings: [{unitName:'조각',gramsPerUnit:130},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Kimchi', name: '김치', category: 'korean', subcategory: 'side', servings: [{unitName:'접시',gramsPerUnit:80},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Seaweed, dried', name: '김', category: 'korean', subcategory: 'side', servings: [{unitName:'장',gramsPerUnit:3},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Cheese, cheddar', name: '체다치즈', category: 'western', subcategory: 'side', servings: [{unitName:'조각',gramsPerUnit:20},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Yogurt, plain, low fat', name: '플레인요거트', category: 'western', subcategory: 'side', servings: [{unitName:'컵',gramsPerUnit:245},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Milk, whole', name: '우유', category: 'western', subcategory: 'side', servings: [{unitName:'컵',gramsPerUnit:240},{unitName:'ml',gramsPerUnit:1}] },
  { query: 'Banana, raw', name: '바나나', category: 'western', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:120},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Apple, raw', name: '사과', category: 'western', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:180},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Soy sauce', name: '간장', category: 'korean', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:15},{unitName:'ml',gramsPerUnit:1}] },
  { query: 'Butter, salted', name: '버터', category: 'western', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:14},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Oil, olive', name: '올리브유', category: 'western', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:14},{unitName:'ml',gramsPerUnit:1}] },
  { query: 'Peanut butter, smooth style', name: '땅콩버터', category: 'western', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:16},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Honey', name: '꿀', category: 'western', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:21},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Sugar, granulated', name: '설탕', category: 'western', subcategory: 'side', servings: [{unitName:'T',gramsPerUnit:12},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Chocolate, dark', name: '다크초콜릿', category: 'western', subcategory: 'snack', servings: [{unitName:'조각',gramsPerUnit:10},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Ice cream, vanilla', name: '바닐라아이스크림', category: 'western', subcategory: 'snack', servings: [{unitName:'스쿱',gramsPerUnit:70},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Sweet potato, cooked, baked', name: '고구마', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:200},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Corn, sweet, yellow, cooked, boiled', name: '옥수수', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Potatoes, boiled, cooked without skin, flesh', name: '감자', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Onions, raw', name: '양파', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:150},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Garlic, raw', name: '마늘', category: 'korean', subcategory: 'side', servings: [{unitName:'쪽',gramsPerUnit:5},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Carrots, raw', name: '당근', category: 'korean', subcategory: 'side', servings: [{unitName:'개',gramsPerUnit:70},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Spinach, raw', name: '시금치', category: 'korean', subcategory: 'side', servings: [{unitName:'접시',gramsPerUnit:80},{unitName:'g',gramsPerUnit:1}] },
  { query: 'Cabbage, raw', name: '양배추', category: 'korean', subcategory: 'side', servings: [{unitName:'접시',gramsPerUnit:100},{unitName:'g',gramsPerUnit:1}] },
]

async function main() {
  const results = [];
  for (const f of FOODS) {
    process.stderr.write(`Fetching: ${f.name}... `);
    const data = await fetchUSDA(f.query);
    if (data && data.energy) {
      results.push({
        ...f,
        caloriesPer100g: Math.round(data.energy * 10) / 10,
        proteinPer100g: Math.round((data.protein || 0) * 10) / 10,
        fatPer100g: Math.round((data.fat || 0) * 10) / 10,
        carbsPer100g: Math.round((data.carbs || 0) * 10) / 10,
        sourceName: data.name,
      });
      process.stderr.write(`✅ ${data.energy}kcal | P:${data.protein}g F:${data.fat}g C:${data.carbs}g\n`);
    } else {
      process.stderr.write(`❌\n`);
    }
    await new Promise(r => setTimeout(r, 200));
  }
  console.log(JSON.stringify(results, null, 2));
  process.stderr.write(`\nDone! ${results.length}/${FOODS.length}\n`);
}

main();
